[c10d] Flush file in file recorder

c-p-i-o opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145458

Summary:
Flushing file to hopefully prevent file corruptions as reported in
https://github.com/pytorch/pytorch/pull/145125

Test Plan:
Couldn't get file corruption to occur in my tests.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

c-p-i-o committed:
[c10d] Flush file in file recorder

Summary:
Flushing file to hopefully prevent file corruptions as reported in
https://github.com/pytorch/pytorch/pull/145125

Test Plan:
Couldn't get file corruption to occur in my tests.

[ghstack-poisoned]

c-p-i-o committed:
Update on "[c10d] Flush file in file recorder"


Summary:
Flushing file to hopefully prevent file corruptions as reported in
https://github.com/pytorch/pytorch/pull/145125

Test Plan:
Couldn't get file corruption to occur in my tests.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

c-p-i-o committed:
Update on "[c10d] Flush file in file recorder"


Summary:
Flushing file to hopefully prevent file corruptions as reported in
https://github.com/pytorch/pytorch/pull/145125

Test Plan:
Couldn't get file corruption to occur in my tests.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -159,18 +159,31 @@ void DebugInfoWriter::write(const std::string& trace) {
-  if (!file) {
-    LOG(ERROR) << "Error opening file for writing NCCLPG debug info: "
+  if (!file.write(trace.data(), static_cast<std::streamsize>(trace.size()))) {
+    if (file.bad()) {
+      LOG(ERROR) << "Serious error writing Flight Recorder debug info to file: "
+                 << filename_;
+    } else {
+      LOG(ERROR) << "Error writing Flight Recorder debug info to file: "
+                 << filename_;
+    }
```
Curious, what is the distinction between "serious error" and error?

c-p-i-o replied:
I just just leave it as a single error - less code. Let me clean this up.
From the documentation - bad usually means something is seriously wrong with the file.  
https://cplusplus.com/reference/ios/ios/bad/

kwen2501 reviewed (approved):
LGTM.

c-p-i-o committed:
Update on "[c10d] Flush file in file recorder"


Summary:
Flushing file to hopefully prevent file corruptions as reported in
https://github.com/pytorch/pytorch/pull/145125

Test Plan:
Couldn't get file corruption to occur in my tests.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

c-p-i-o commented:
@pytorchbot merge

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o commented:
@pytorchbot merge

c-p-i-o commented:
@pytorchbot merge
